### PR TITLE
Tutorial: Reactor Event - AWS Lambda

### PR DESCRIPTION
**Pull request by axefrog**

### CODE REVIEW

File: content/tutorials/reactor-event-aws.textile [hunk 1]
```diff
@@ -0,0 +1,785 @@
+---
+title: Reactor Events Tutorial: AWS Lambda Function
```
[the remark this answers is not in the document]
Comment (axefrog) [reply content/tutorials/reactor-event-aws.textile]: Acknowledged, fixed in next PR.

File: content/tutorials/reactor-event-aws.textile [hunk 2]
```diff
@@ -0,0 +1,785 @@
...
+section: tutorials
+index: 100
+languages:
+  - javascript
+  - nodejs
+---
+
+h1. Reactor Events Tutorial: AWS Lambda Function
```
[the remark this answers is not in the document]
Comment (axefrog) [reply content/tutorials/reactor-event-aws.textile]: Acknowledged, fixed in next PR.

File: content/tutorials/reactor-event-aws.textile [hunk 3]
```diff
@@ -0,0 +1,785 @@
...
+
+The "Ably Realtime":https://www.ably.io/ service provides a "pub/sub" (publish/subscribe) messaging system with which you can build sophisticated workflows and dataflow pipelines to meet the needs of your application. This can be as simple as publishing messages to named channels of your choice, and having users and server processes subscribe to those channels and respond with further messages as needed. But what if you don't have a server to coordinate all of this, or simply don't want to manage that side of things?
```
[the remark this answers is not in the document]
Comment (axefrog) [reply content/tutorials/reactor-event-aws.textile]: Acknowledged, fixed in next PR.

File: content/tutorials/reactor-event-aws.textile [hunk 4]
```diff
@@ -0,0 +1,785 @@
...
+
+h2. What we'll be building
+
+For this tutorial, we'll be using *"Reactor Events":https://www.ably.io/reactor* in response to incoming messages in order to *automatically invoke your own AWS Lambda functions*. We're going to build a simple pizza ordering system with a virtual pizza assistant chat bot. When a new customer hits your website, they'll be assigned a random ID, their messages will be published to a private channel created just for that customer, and a Reactor Event will call your AWS Lambda functions to invoke the pizza chat bot, who will take their order and then, via an HTTPS request to the Ably REST API, post a response back to a similar channel that the web application will use to listen for responses from the chat bot.
```
[the remark this answers is not in the document]
Comment (axefrog) [reply content/tutorials/reactor-event-aws.textile]: Ok, as I apply the remaining requested revisions I'll keep an eye out for any other sentences that are more complex than they need to be, so keep an eye out for those in the next PR.

File: content/tutorials/reactor-event-aws.textile [hunk 5]
```diff
@@ -0,0 +1,785 @@
...
+
+Enter a name such as "Lambda-Lex-Role", followed by a description of the role's purpose, and then click the "Create Role" button:
+
+<a href="/images/tutorials/reactor-event-aws-iam-5.png" target="_blank">
+  <img src="/images/tutorials/reactor-event-aws-iam-5.png" style="width: 100%" alt="Fill out the role details">
+</a>
+
+--------------------------------------------------------------------------------
+
+h3. (c) IAM user for use by your Ably Reactor event
```
[the remark this answers is not in the document]
Comment (axefrog) [reply content/tutorials/reactor-event-aws.textile]: Acknowledged, fixed in next PR.

File: content/tutorials/reactor-event-aws.textile [hunk 6]
```diff
@@ -0,0 +1,785 @@
...
+
+--------------------------------------------------------------------------------
+
+h2. Step 2: AWS Lex
+
+bq. _Note: You can skip this step if you're not going to use AWS Lex for your chat bot._
+
+h3. (a) Python and the AWS CLI
+
+Make sure you have Python installed. A Python 3.6.x installation is used in this tutorial. You can "get Python here":https://www.python.org/downloads/, and this will also give you the *pip* installer, which you'll need to install the AWS CLI. If you're familiar with Node.js, pip is a package manager like npm, but for Python.
```
[the remark this answers is not in the document]
Comment (axefrog) [reply content/tutorials/reactor-event-aws.textile]: I've shortened this down as requested, though I retained a modified version of the tail section as it references the requirements relevant to the tutorial. I can kill it entirely though, if necessary.

File: content/tutorials/reactor-event-aws.textile [hunk 7]
```diff
@@ -0,0 +1,785 @@
...
+
+Open your command line shell of choice. This would likely be Power Shell or Bash (WSL) on Windows, or a Bash terminal on Mac and Linux. Check that Python and pip are available by typing `python -V` to print out the installed version of Python, and `pip -V` to check that pip is also installed.
```
[the remark this answers is not in the document]
Comment (axefrog) [reply content/tutorials/reactor-event-aws.textile]: These parts are gone now, as per the previous comment.

File: content/tutorials/reactor-event-aws.textile [hunk 8]
```diff
@@ -0,0 +1,785 @@
...
+
+As per the instructions on "this page":http://docs.aws.amazon.com/cli/latest/userguide/installing.html, install the AWS CLI by typing `pip install awscli`. This will download and install the AWS CLI as a global Python package that you can access from the command line. Once you've done this, type `aws configure`. You'll be prompted to add the credentials you retrieved earlier. You should also choose *us-east-1* for the region (AWS Lex isn't available in all regions), and *json* for the output format.
```
[the remark this answers is not in the document]
Comment (axefrog) [reply content/tutorials/reactor-event-aws.textile]: As above.

File: content/tutorials/reactor-event-aws.textile [hunk 9]
```diff
@@ -0,0 +1,785 @@
...
+
+The JSON configurations for a simple pizza assistant chat bot are included in "the repository for this tutorial's completed code":http://todo-repository-link.
+
+*This preconfigured bot is only a bare-bones implementation to demonstrate a proof of concept. Customising the bot's behaviour and filling in the gaps, such that it is able to actually complete an order or manage a list of different toppings in a robust manner, is left as an exercise for the reader.*
+
+From your console or terminal, change to the directory where you've put the JSON files, and run the following:
+
+```[sh]
+aws lex-models put-slot-type ---cli-input-json file://slottype-crust.json
```
[the remark this answers is not in the document]
Comment (axefrog) [reply content/tutorials/reactor-event-aws.textile]: Acknowledged, fixed in next PR.

File: content/tutorials/reactor-event-aws.textile [hunk 10]
```diff
@@ -0,0 +1,785 @@
...
+aws lex-models put-slot-type ---cli-input-json file://slottype-sauce.json
+aws lex-models put-slot-type ---cli-input-json file://slottype-size.json
+aws lex-models put-slot-type ---cli-input-json file://slottype-topping.json
+aws lex-models put-intent ---cli-input-json file://intent.json
+aws lex-models put-bot ---cli-input-json file://bot.json
+aws lex-models put-bot-alias ---cli-input-json file://alias.json
+```
+
+This will create and configure your chat bot so that it can take a simple pizza order via a text-based conversation with a user. You'll use your Lambda function to route user messages from your Ably channel to Lex, and to post the bot's response back to the user on the same channel.
```
[the remark this answers is not in the document]
Comment (axefrog) [reply content/tutorials/reactor-event-aws.textile]: Acknowledged, fixed in next PR.

File: content/tutorials/reactor-event-aws.textile [hunk 11]
```diff
@@ -0,0 +1,785 @@
...
+
+--------------------------------------------------------------------------------
```
[the remark this answers is not in the document]
Comment (axefrog) [reply content/tutorials/reactor-event-aws.textile]: Acknowledged, all horizontal rules removed in next PR.

File: content/tutorials/reactor-event-aws.textile [hunk 12]
```diff
@@ -0,0 +1,785 @@
...
+
+Fill out the form to provide Ably with the details it needs to invoke your Lambda function.
+
+# Select `us-east-1` for the region.
+# Enter `Invoke-PizzaAssistant` - the name of your Lambda function.
+# Select the "AWS Credentials" option.
+# The key ID and secret key you'll be using are for the IAM user you created in step 1-c of this tutorial. The key ID and secret key should be pasted one after the other with a colon in-between. For example if your key ID is `foo` and your secret key is `bar`, you'd enter `foo:bar` into the credentials field.
+# Make sure "Message" is set as the source.
+# Enter a regular expression that the Reactor event will use to whitelist the Ably channels for which the event should be triggered. `^pizza:customer:` means that the channel name must begin with the string `pizza:customer:`, with no restrictions on the characters that follow.
+# The JSON option means that your channel uses standard JSON packets for messaging, rather than custom-formatted data, as might be the case if your channel were being used for voice or other binary data.
```
[the remark this answers is not in the document]
Comment (axefrog) [reply content/tutorials/reactor-event-aws.textile]: Acknowledged, fixed in next PR.

[the remark this answers is not in the document]
Comment (axefrog) [reply content/tutorials/reactor-event-aws.textile]: Is the "Enveloped" checkbox to be excluded as well?

[the remark this answers is not in the document]
Comment (axefrog) [reply content/tutorials/reactor-event-aws.textile]: I've left it in, but can remove it from the screenshot if it was related to the JSON option.

File: content/tutorials/reactor-event-aws.textile [hunk 13]
```diff
@@ -0,0 +1,785 @@
...
+
+```[js]
+// This Ably client library is available due to the script in the HTML header:
+const ably = Ably.Realtime('YOUR_API_KEY_GOES_HERE');
+
+// Remember when creating your API key that you specified that it would only
+// work for channels starting with the string "pizza:". When testing the script,
+// try changing the channel name to something else and see how Ably emits an
+// error to your browser developer console, thus demonstrating the safety of
+// having the API key in a client-side browser script.
```
[the remark this answers is not in the document]
Comment (axefrog) [reply content/tutorials/reactor-event-aws.textile]: Acknowledged, comma and what follows are removed in the next PR.

File: content/tutorials/reactor-event-aws.textile [hunk 14]
```diff
@@ -0,0 +1,785 @@
...
+
+h3. (c) Filling out the primary functionality
+
+Now that you've verified that you're able to connect to Ably and send and receive messages, it's time to update the code to integrate it with everything you did earlier in this tutorial.
+
+First, we'll need two channels rather than one. The first channel is the outbound channel, and is where the Reactor event will be listening for messages to forward on to your AWS Lambda function. The second channel is the inbound channel where the bot's response messages will be posted, and is the channel to which your script will attach a listener.
+
+Also, because channels are lightweight, it's ok to make as many channels as you need. If two different customers are trying to order a pizza at the same time, we don't want their messages getting mixed in together, so we'll be appending a random customer identification value to the inbound and outbound channel names, with the effect being that every customer is allocated their own unique pair of inbound and outbound channels. The customer ID will also need to be included in outbound messages so that your AWS Lambda function can figure out where to send responses.
```
[the remark this answers is not in the document]
Comment (axefrog) [reply content/tutorials/reactor-event-aws.textile]: Acknowledged, fixed in next PR.

File: content/tutorials/reactor-event-aws.textile [hunk 15]
```diff
@@ -0,0 +1,785 @@
...
+const outboundChannel = ably.channels.get('pizza:customer:' + customerId);
+
+// Messages from the bot will be posted here, where we'll be listening for responses
+const inboundChannel = ably.channels.get('pizza:bot:' + customerId);
+
+// Update your subscription to listen to the inbound channel
+inboundChannel.subscribe(receiveMessage);
+```
+
+Let's refactor the message publishing code into its own function. We'll also need to modify what we're sending so that it includes the customer ID so that the Lambda function can reply on the correct channel:
```
[the remark this answers is not in the document]
Comment (axefrog) [reply content/tutorials/reactor-event-aws.textile]: I'll reword this to emphasize convenience over necessity.

File: content/tutorials/reactor-event-aws.textile [hunk 16]
```diff
@@ -0,0 +1,785 @@
...
+  // As above, but trap errors that prevent the call from happening at all:
+  req.on('error', (err) => {
+    callback(err);
+  });
+
+  // Send the payload, and we're done:
+  req.write(data);
+  req.end();
+}
+```
```
[the remark this answers is not in the document]
Comment (axefrog) [reply content/tutorials/reactor-event-aws.textile]: Sure, but it seems the user would have to create a serverless deploy package for this (correct me if I'm wrong on that account). The tutorial just uses the AWS UI for the tutorial, so the code uses the REST API on that basis.

File: content/tutorials/reactor-event-aws.textile (hunk 12 above)
[the remark this answers is not in the document]
Comment (axefrog) [reply content/tutorials/reactor-event-aws.textile]: Disregard, I see you've indirectly answered this question further down in the comments.

File: content/tutorials/reactor-event-aws.textile (hunk 16 above)
[the remark this answers is not in the document]
Comment (axefrog) [reply content/tutorials/reactor-event-aws.textile]: Note that I made the following comment on the Slack channel:

> It may end up being worthwhile updating the main reactor event AWS tutorial to use the new chat bot script I wrote, instead of Lex (seeing as the new script is a lot more robust). I'd prefer to do this as a separate job though, if you decide that's what you want.

As an extension to the above, I could re-focus the tutorial on using a serverless deploy package instead.

File: content/tutorials/reactor-event-aws.textile [hunk 17]
```diff
@@ -0,0 +1,785 @@
...
+
+One last step before leaving the AWS control panel - it'd be good if we could quickly test the updated Lambda function to make sure it's working before proceeding. Update the test configuration you originally created, replacing the short JSON configuration you had earlier with the JSON in the code block below ("also in the tutorial repository":http://todo-repository-link/lambda-test.json). This JSON data reflects the fact that Ably delivers the customer's message wrapped in some additional metadata, which is why we need to unwrap the payload (note the `JSON.parse` line in the script block above) to get access to the data we want.
```
[the remark this answers is not in the document]
Comment (axefrog) [reply content/tutorials/reactor-event-aws.textile]: I've added a note to this effect.

File: content/tutorials/reactor-event-aws.textile [hunk 18]
```diff
@@ -0,0 +1,785 @@
...
+focusInputField();
+```
+
+Congratulations, you're done!
+
+<a href="/images/tutorials/reactor-event-aws-project-animation.gif" target="_blank">
+  <img src="/images/tutorials/reactor-event-aws-project-animation.gif" style="width: 100%" alt="Animated image of the browser app in action">
+</a>
+
+h2. Next Steps
```
[the remark this answers is not in the document]
Comment (axefrog) [reply content/tutorials/reactor-event-aws.textile]: Acknowledged, fixed in next PR.

**Comment by axefrog** [comment]
I have tried to build the docs both via docker and ruby, but I'm roadblocked by connection failures to sourceforge in both cases, while trying to complete the toolchain setup. For what it's worth, I'm on Windows - I don't own a Mac.

In any case, I've fixed all the backticks; let me know if you can see any other minor things I can fix, but which aren't obvious without building the docs myself.

File: content/tutorials/reactor-event-aws.textile (hunk 16 above)
[the remark this answers is not in the document]
Comment (SimonWoolf) [reply content/tutorials/reactor-event-aws.textile]: > Sure, but it seems the user would have to create a serverless deploy package for this (correct me if I'm wrong on that account). The tutorial just uses the AWS UI for the tutorial, so the code manually calls the REST API on that basis

Thing is, we want to seriously discourage people from using the REST api manually if at all possible, in favour of the client libs, for a number of reasons (e.g. automatic retries to fallback hosts to route around region-specific or DNS issues). As such we would like all the tutorials to use the official client libs, even if it's a bit of extra work.

You do have to create a package, but you only need a pretty minimal one. I was actually thinking of making an official example lambda package anyway. Tell you what -- I'll do that now, so you can link to it in the tutorial, to save the trouble of explaining how to create the package from scratch. so something "Clone this example package [to be done], replace the contents of index.js with the following code, and run `npm run zip` to generate the zip package".

**Comment by SimonWoolf** [comment]
> I'm roadblocked by connection failures to sourceforge in both cases, while trying to complete the toolchain setup. For what it's worth, I'm on Windows - I don't own a Mac.

For windows, the easiest way is probably to use the windows linux subsystem, see https://docs.microsoft.com/en-us/windows/wsl/install-win10 (the 'ubuntu' path). once that's done, you can install ruby and git with `sudo apt-get install ruby bundler git`, then from a bash shell, git-clone the docs repo, and you should then be able to follow the steps in https://github.com/ably/docs#forking-and-running-locally .

File: content/tutorials/reactor-event-aws.textile (hunk 16 above)
[the remark this answers is not in the document]
Comment (SimonWoolf) [reply content/tutorials/reactor-event-aws.textile]: Example lambda package up at https://github.com/ably/example-lambda-function
